### PR TITLE
Make font size an implicit attribute of the block

### PR DESCRIPTION
**Pull request by oandregal**
This PR follows suit on what was done for colors at https://github.com/WordPress/gutenberg/pull/21021 and makes the font size an implicit attribute of the block. Implemented for the paragraph block.

**Notes**

Custom font sizes might not be applied properly depending on the theme styles. If themes apply editor styles like this:

```
p {
   font-size: 12px
}
```

We'll recommend themes to do something like this instead

```
p { font-size: var( --wp--typography--font-size, 12px ); }
```

### CODE REVIEW

File: packages/block-editor/src/hooks/font-size.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+		if ( ! hasBlockSupport( blockName, FONT_SIZE_SUPPORT_KEY ) ) {
+			return <BlockEdit key="edit" { ...props } />;
+		}
+
+		return [
+			<FontSizeInspectorControl
+				key="inspector-controls"
+				fontSize={ fontSize }
+				setFontSize={ setFontSize }
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/font-size.js]: should we set the font size to the "style" attribute instead when it's a custom font size (like the colors)

File: packages/block-editor/src/hooks/font-size.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+function addSaveProps( props, blockType, attributes ) {
+	if ( ! hasBlockSupport( blockType, FONT_SIZE_SUPPORT_KEY ) ) {
+		return props;
+	}
+
+	const { fontSize } = attributes;
+
+	const fontSizeClass = getFontSizeClass( fontSize );
+
+	props.className = classnames( props.className, fontSizeClass );
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/font-size.js]: Sometimes this results in an empty string we should return undefined in these cases (I had the issue in the color hook and fixed it in master)

File: packages/block-editor/src/hooks/font-size.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+		}
+		return addSaveProps( props, settings, attributes );
+	};
+
+	return settings;
+}
+
+const FontSizeInspectorControl = ( { value, onChange } ) => (
+	<InspectorControls key="inspector-controls">
+		<PanelBody title={ __( 'Text settings' ) }>
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/font-size.js]: In #20775 we're adding the line height hook into a new "Typography" panel. These two hooks need to be added to the same panel so we need to sync when one of the PR merges.

File: packages/block-editor/src/hooks/font-size.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+			<FontSizePicker value={ value } onChange={ onChange } />
+		</PanelBody>
+	</InspectorControls>
+);
+
+const setFontSize = ( { style, setAttributes }, fontSizes ) => ( value ) => {
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/font-size.js]: Nit: This doesn't seem to bring much value as a separate function, it makes things a bit less readable to me, I think it's better as a function inside the component.

**Review by youknowriad — COMMENTED** [review]
Nice work here, seems like we're closed. I think we might need a deprecated version to migrate the existing customFontSize attribute. I think you can just reuse the last deprecated version (since it's not released yet and was just added) and add the font size migration (similar to the color one)

File: packages/block-editor/src/hooks/font-size.js (hunk 1 above)
Comment (nosolosw) [reply packages/block-editor/src/hooks/font-size.js]: Yes! I just submitted what I had at the time, still need finishing some things.

File: packages/block-editor/src/hooks/font-size.js (hunk 3 above)
Comment (nosolosw) [reply packages/block-editor/src/hooks/font-size.js]: It looks like that's ready to merge. I'd let that get in first and then we can work on this and discuss how to merge things.

I was following the thread and my current thoughts are:

- In terms of how blocks would opt-in =>
  - font settings are one thing: line height, font size, font weight, font family (you don't opt-in to each of them separately).
  - drop cap is a separate support key.
- In terms of UI => 
  - All font settings (font size, weight, line-height, font family) would be enabled at once if you opt-in into font and will use the same panel.
  - The drop cap should also use the same panel but will be enabled separately.

How does this sound?

cc @ItsJonQ

Comment (youknowriad) [reply packages/block-editor/src/hooks/font-size.js]: I don't have a precise answer on the opt-in flag. Maybe all these flags are part of the same "style" flag but I think at this moment, it's too soon to make the best decision. We should just continue with experimental flags and adjust when we want to stabilize the approach.

In terms of UI, it makes sense to gather all typography settings IMO. Potentially, each hook would export a "component" or something like that and everything is applied on the same "style" hook.

Comment (ItsJonQ) [reply packages/block-editor/src/hooks/font-size.js]: > In terms of how blocks would opt-in... In terms of UI ...

@nosolosw That makes sense to me! Attributes like `line-height`, `font-weight`, etc... are very generic. As a user, I would expect them to be available anywhere I could adjust text.

Drop cap on the other hand... is very specific.

Based on that model, I think keeping it separate makes sense!

<3

Comment (youknowriad) [reply packages/block-editor/src/hooks/font-size.js]: Looks like the other PR is merged, time to rebase and see how we can merge things :)

File: packages/block-editor/src/hooks/font-size.js [hunk 5]
```diff
@@ -60,10 +61,11 @@ function addSaveProps( props, blockType, attributes ) {
 		return props;
 	}
 
-	const { fontSize } = attributes;
+	// Use TokenList to dedupe classes.
+	const classes = new TokenList( props.className );
+	classes.add( getFontSizeClass( attributes.fontSize ) );
+	const newClassName = classnames( classes.value );
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/font-size.js:67]: do we really need `classnames` call now?

Comment (nosolosw) [reply packages/block-editor/src/hooks/font-size.js]: nah, we don't https://github.com/WordPress/gutenberg/pull/21153/commits/652e85b9b22e932bdf17d36ef5639da88e1180b4

**Review by youknowriad — APPROVED** [review]
LGTM 👍

File: packages/block-editor/src/hooks/font-size.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+function addSaveProps( props, blockType, attributes ) {
+	if ( ! hasBlockSupport( blockType, FONT_SIZE_SUPPORT_KEY ) ) {
+		return props;
+	}
+
+	// Use TokenList to dedupe classes.
+	const classes = new TokenList( props.className );
+	classes.add( getFontSizeClass( attributes.fontSize ) );
+	const newClassName = classes.value;
+	props.className = newClassName ? newClassName : undefined;
```
Comment (nosolosw) [inline packages/block-editor/src/hooks/font-size.js:63]: Noting here that this has the side-effect of removing duplicated classes for old blocks that are updated. I can't think of any reason why this would be a problem but thought to highlight anyway before merging. cc @youknowriad 

I've testing this with old content and works nicely, so I'm going to land in a few hours.

**Comment by nosolosw** [comment]
This needs fixing an e2e snapshot. Have to go AFK for a couple of hours, will fix later.

File: packages/block-library/src/style.scss [hunk 7]
```diff
@@ -217,28 +217,30 @@
 
 
 // Font sizes.
+:root {
```
Comment (nosolosw) [inline packages/block-library/src/style.scss:220]: Ideally, this could use the CSS variables as well, can do in a separate PR. cc @jorgefilipecosta as you were thinking about this.

Comment (nosolosw) [reply packages/block-library/src/style.scss:220]: Follow-up at https://github.com/WordPress/gutenberg/pull/21429

**Comment by kjellr** [comment]
> Custom font sizes might not be applied properly depending on the theme styles.

Sorry — I didn't see this when I was pinged on Friday... This (and the other recently merged `line-height` PR is breaking some in-progress theme work at the moment, so I'd love to get some clarity on how themes are supposed to work with this. 

At the moment, the `:root` specificity of this rule (and the line height one) tends to override things far beyond what's intended. For instance, in Twenty Nineteen, the site title and site description are both `p` tags on post pages. The `line-height` and `font-size` for those is now being overwritten by this rule: 

twenty-nineteen

The recommended note above is not designed to solve this problem: 

```
p { font-size: var( --wp--typography--font-size, 12px ); }
```

The issue here is the `:root` specificity. So the theme would be required to use selectors like this to override in this case: 

```
:root .site-title, 
:root .site-description
```

... and that seems a little weird. Is there a way for just to implement these new variables without relying on that `:root` specificity?

**Comment by youknowriad** [comment]
@kjellr Thanks for the comment and indeed the Global styles work has potential to conflict with theme styles.

We're still figuring out the best recommendations for theme authors and in the meantime we'll have an opt-in way to switch to the new styles. See #21428 

That said, the solutions for theme will most likely involve:
 - writing less CSS
 - Adding config to the `theme.json` file
 - Use CSS variables to set these colors/customizations instead of raw values. 

Can you clarify why theme would be required to do this `:root .site-title` but this instead?

```
.site-title {
    font-size: 35px;
     --wp--typography--font-size: 35px;
}

**Comment by kjellr** [comment]
> We're still figuring out the best recommendations for theme authors and in the meantime we'll have an opt-in way to switch to the new styles. See #21428

Cool, that'll be necessary. Right now, these styles override _lots_ of existing themes' styles. 

> Can you clarify why theme would be required to do this :root .site-title but this instead?

I see what you're saying. Something like that would work, but it doesn't feel very intuitive to me at first. I usually find that the variables are defined solely in the `root`, and then modified without actually redefining the value (eg. `font-size: (var(----wp--typography--font-size) * 2)`). This is likely just habit, influenced by many years of using Sass variables though. I can see how in a totally new theme (one built for global styles), this could be a relatively easy approach though.
